pkg/vendir/fetch/hg/hg.go

```diff
@@ -0,0 +1,258 @@
+	if err != nil {
+		return HgInfo{}, err
+	}
+
+	info := HgInfo{}
+
+	out, _, err := t.run([]string{"id", "-i"}, nil, dstPath)
+	if err != nil {
+		return HgInfo{}, err
+	}
+
+	info.SHA = strings.TrimSpace(out)
+
+	out, _, err = t.run([]string{"log", "-l", "1", "-T", "{desc|firstline|strip}", "-r" , info.SHA}, nil, dstPath)
+	if err != nil {
+		return HgInfo{}, err
+	}
+
+	info.CommitTitle = strings.TrimSpace(out)
+
+	return info, nil
+}
+
+func (t *Hg) fetch(dstPath string, tempArea ctlfetch.TempArea) error {
+	authOpts, err := t.getAuthOpts()
+	if err != nil {
+		return err
+	}
+
+	authDir, err := tempArea.NewTempDir("hg-auth")
+	fmt.Printf(">>> authDir: %s", authDir)
+	if err != nil {
+		return err
+	}
+
+	defer os.RemoveAll(authDir)
+
+	env := os.Environ()
+
+	hgURL := t.opts.URL
+
+	_, _, err = t.run([]string{"init"}, env, dstPath)
+	if err != nil {
+		return err
+	}
+
+	hgrcPath := filepath.Join(dstPath, ".hg", "hgrc")
+
+	hgrc := fmt.Sprintf(`[paths]
+default = %s
+`, hgURL)
+
+	pullCmd := []string{"pull"}
+
+	if authOpts.Username != nil && authOpts.Password != nil {
+		if !strings.HasPrefix(hgURL, "https://") {
+			return fmt.Errorf("Username/password authentication is only supported for https remotes")
+		}
+		hgCredsURL, err := url.Parse(hgURL)
+		if err != nil {
+			return fmt.Errorf("Parsing hg remote url: %s", err)
+		}
+
+		pullCmd = []string{
+			"pull",
+			fmt.Sprintf("auth.hgauth.prefix = https://%s", hgCredsURL.Host),
+			fmt.Sprintf("auth.hgauth.username = %s", *authOpts.Username),
+			fmt.Sprintf("auth.hgauth.password = %s", *authOpts.Password),
+		}
+
+	}
+
+	if authOpts.IsPresent() {
+		sshCmd := []string{"ssh", "-o", "ServerAliveInterval=30", "-o", "ForwardAgent=no", "-F", "/dev/null"}
+
+		if authOpts.PrivateKey != nil {
+			path := filepath.Join(authDir, "private-key")
+
+			err = ioutil.WriteFile(path, []byte(*authOpts.PrivateKey), 0600)
+			if err != nil {
+				return fmt.Errorf("Writing private key: %s", err)
+			}
+
+			sshCmd = append(sshCmd, "-i", path, "-o", "IdentitiesOnly=yes")
+		}
+
+		if authOpts.KnownHosts != nil {
+			path := filepath.Join(authDir, "known-hosts")
+
+			err = ioutil.WriteFile(path, []byte(*authOpts.KnownHosts), 0600)
+			if err != nil {
+				return fmt.Errorf("Writing known hosts: %s", err)
+			}
+
+			sshCmd = append(sshCmd, "-o", "StrictHostKeyChecking=yes", "-o", "UserKnownHostsFile="+path)
+		} else {
+			sshCmd = append(sshCmd, "-o", "StrictHostKeyChecking=no")
+		}
+
+		hgrc = fmt.Sprintf(`%s
+[ui]
+ssh = %s
+`, hgrc, strings.Join(sshCmd, " "))
+	}
+
+	err = ioutil.WriteFile(hgrcPath, []byte(hgrc), 0600)
+	if err != nil {
+		return fmt.Errorf("Writing %s: %s", hgrcPath, err)
+	}
+
+	_, _, err = t.run(pullCmd, env, dstPath)
+	if err != nil {
+		return err
+	}
+
+	ref, err := t.resolveRef(dstPath)
+	if err != nil {
+		return err
+	}
+
+	_, _, err = t.run([]string{"checkout", ref}, env, dstPath)
+
+	return err
+}
+
+func (t *Hg) resolveRef(dstPath string) (string, error) {
```

Comment:
doesnt look like this is used. in git impl we are grabbing specific tag before checking out.

Reply:
yes, like I said in my previous comment.

Reply:
didnt a see a comment, but saw you removed it.